NOTICE.txt

```diff
@@ -0,0 +1,7 @@
+==========
+Notice for: elastic-apm-mongo
+----------
+Based on the elastic-apm-mongo project by Vadim Hatsura (@vhatsura),
+licensed under the Apache 2.0 License. https://github.com/vhatsura/elastic-apm-mongo
+
+<text of the Apache 2.0 License>
```

Comment:
The NOTICE.txt should be added to the Elastic.Apm.MongoDb .csproj in order to pack it into the Nuget package distribution, similar to how the shared LICENSE file is packed in src/Directory.Build.Props

https://github.com/elastic/apm-agent-dotnet/blob/26ebf44b49446f1b5b2e79a6724bc8502d1ca5d2/src/Directory.Build.props#L33-L36

What's not clear to me is whether NOTICE.txt should be included in _all_ Nuget packages - @gtback? 

I _think_ we can also remove `<text of the Apache 2.0 License>`, based on [Apache's example NOTICE file](https://infra.apache.org/licensing-howto.html#example-notice). @gtback?

Reply:
minor nitpick: Might be good to use the file name NOTICE, for consistency with the existing LICENSE in solution root.

Reply:
> What's not clear to me is whether NOTICE.txt should be included in _all_ Nuget packages - @gtback?

I'm not sure exactly how Nuget works, but I think you mean all of the various packages [here](https://www.nuget.org/packages?q=elastic.apm) (minus the ones we don't publish, of course). Any package which includes the MongoDB code should definitely have it, but if there are any packages that only cause the MongoDB code to be separately downloaded, I don't think we need it (but it wouldn't hurt). Does that make sense?
 
> I _think_ we can also remove `<text of the Apache 2.0 License>`, based on [Apache's example NOTICE file](https://infra.apache.org/licensing-howto.html#example-notice). @gtback?

Our policy is to include the full text of any licenses of any dependencies listed in NOTICE.txt. I meant to put the actual full text of the license here, not the literal text `<text of the Apache 2.0 License>`; sorry for any confusion.

That said, given that the entire project is already Apache-2.0 licensed, I think we could get rid of this without causing any problems (we don't need the same text in both `LICENSE.txt` and `NOTICE.txt`

> minor nitpick: Might be good to use the file name NOTICE, for consistency with the existing LICENSE in solution root.

Either is fine. I think consistency within a repo is more important than consistency between repos.

Reply:
Thanks for the input on this.

1. > Might be good to use the file name NOTICE

Agreed - renamed it and merged it with what we already had.

2. Regarding the `<text of the Apache 2.0 License>`: 

I removed it and I did not add the Apache 2.0 text itself.

3. > What's not clear to me is whether NOTICE.txt should be included in all Nuget packages 

We already have one notice for `Elastic.Apm` and now we have one for `Elastic.Apm.MongoDb`. I think the best and easiest would be to include this in all our packages. 2 advantages I see here: 1) no matter which package people use, they'll be able to easily see our notices - otherwise they'd need to collect the info from multiple NuGet packages following dependencies, 2) it's easier for us to maintain (single file, we include the same in all packages).

So what I did is that I added the NOTICE file to `apm-agent-dotnet/src/Directory.Build.props` (same as our LICENSE file), so it'll be included in all our packages.

Any thoughts on this?

Reply:
> So what I did is that I added the NOTICE file to `apm-agent-dotnet/src/Directory.Build.props` (same as our LICENSE file), so it'll be included in all our packages.

I think this is fine.

Reply:
> We already have one notice for Elastic.Apm and now we have one for Elastic.Apm.MongoDb. I think the best and easiest would be to include this in all our packages. 2 advantages I see here: 1) no matter which package people use, they'll be able to easily see our notices - otherwise they'd need to collect the info from multiple NuGet packages following dependencies, 2) it's easier for us to maintain (single file, we include the same in all packages).

> So what I did is that I added the NOTICE file to apm-agent-dotnet/src/Directory.Build.props (same as our LICENSE file), so it'll be included in all our packages.

I am not a lawyer, so would cede to our policies here, but I don't believe that the NOTICE file distributed with a nuget package should include information pertaining to the licenses of derived works that are not part of the nuget package in question. For example,

- The NOTICE file distributed with the Elastic.Apm package should not contain information pertaining to the license of derivative works in Elastic.Apm.MongoDB package

Doing so seems like this would make it much harder for an end consumer of a package to understand what is and isn't relevant, from a compliance perspective. From this point of view, the NOTICE file distributed with a package should only contain information pertaining to the licenses of derived works that are part of the nuget package, whilst the NOTICE file in the root of the source should readily contain information pertaining to the licenses of derivative works in the repository in source form. It would be easier for us to include the same root NOTICE file in every package, and harder to pull out only the relevant bits for each respective package, but in doing the latter, the NOTICE aligns with the actual package contents.

Reply:
Thinking about this more, I agree with @russcam . We generally try to follow the Apache Software Foundation's [practices](https://infra.apache.org/licensing-howto.html) in most cases, and they provide pretty good guidance here.

> The `LICENSE` and `NOTICE` files must exactly represent the contents of the distribution they reside in. Only components and resources that are actually included in a distribution have any bearing on the content of that distribution's `NOTICE` and `LICENSE`.

If it's easier, we can include `NOTICE` files in subdirectories of this repo, if those match up with binary distributions which are published on NuGet.

Reply:
Ok - I moved the two `NOTICE` files into the corresponding projects (aka subdirectories)

We have 2 `NOTICE` files now:
- `Elastic.Apm` has one about `Newtonsoft.json` (`Ben.Demystifier` will be likely added later as well)
- `Elastic.Apm.MondoDb` this is the new one about `vhatsura/elastic-apm-mongo`

Both will be included in the corresponding NuGet package we generate and those packages will only contain notices about the relevant dependencies.